kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/PortConflictDetector.java

```diff
@@ -8,242 +8,114 @@
-     */
-    public void validate(Collection<VirtualClusterModel> virtualClusterModelMap, Optional<HostPort> otherExclusivePort) {
-
-        Set<String> seenVirtualClusters = new HashSet<>();
-
-        Map<Optional<String>, Set<Integer>> inUseExclusivePorts = new HashMap<>();
-        Map<Optional<String>, Map<Integer, Boolean>> inUseSharedPorts = new HashMap<>();
+    private record BindHost(Optional<String> host) {
+        boolean overlaps(BindHost other) {
+            return isAnyHost() || other.isAnyHost() || host.equals(other.host);
+        }
 
-        virtualClusterModelMap.stream()
-                .sorted(Comparator.comparing(VirtualClusterModel::getClusterName))
-                .forEach(virtualCluster -> {
-                    virtualCluster.listeners()
-                            .forEach((listenerName, listener) -> doValidate(otherExclusivePort, listener, inUseExclusivePorts, seenVirtualClusters,
-                                    inUseSharedPorts));
+        boolean isAnyHost() {
+            return host.isEmpty();
+        }
 
-                });
+        @Override
+        public String toString() {
+            return host.orElse(ANY_STRING);
+        }
     }
 
-    private void doValidate(Optional<HostPort> otherExclusivePort, EndpointListener listener, Map<Optional<String>, Set<Integer>> inUseExclusivePorts,
-                            Set<String> seenVirtualClusters, Map<Optional<String>, Map<Integer, Boolean>> inUseSharedPorts) {
-        var proposedSharedPorts = listener.getSharedPorts();
-        var proposedExclusivePorts = listener.getExclusivePorts();
-
-        if (otherExclusivePort.isPresent()) {
-            Optional<String> otherInterface = otherExclusivePort.map(hostPort -> hostPort.host().equals("0.0.0.0") ? null : hostPort.host());
-            var checkPorts = listener.getBindAddress().isEmpty() || otherInterface.isEmpty() || listener.getBindAddress().equals(otherInterface);
-            if (checkPorts) {
-                checkForConflictsWithOtherExclusivePort(otherExclusivePort.get(), listener, proposedExclusivePorts, BindingScope.EXCLUSIVE);
-                checkForConflictsWithOtherExclusivePort(otherExclusivePort.get(), listener, proposedSharedPorts, BindingScope.SHARED);
+    private record CandidateBinding(EndpointListener listener, int port, BindHost bindHost, BindingScope scope) {
+        void validateNonConflicting(CandidateBinding other) {
+            if (this.port == other.port && bindHost.overlaps(other.bindHost)) {
+                if (scope == BindingScope.EXCLUSIVE || other.scope == BindingScope.EXCLUSIVE) {
+                    throw conflictException(other, "exclusive port collision");
+                }
+                if (!bindHost.equals(other.bindHost)) {
+                    throw conflictException(other, "shared port cannot bind to different hosts");
+                }
+                if (listener.isUseTls() != other.listener.isUseTls()) {
```

Comment:
It probably doesn't actually make a practical difference, but I think this is considering the wrong thing.   The condition for shareability ought to be `requireSni == true`.  TLS per-se the deciding factor, it is whether both users of the port agree to use SNI.  If they are not using SNI, there will be no way to distinguish traffic between them.

Reply:
They can't share the port either if one listener wants TLS and the other plain though right?

So I think there's two cases

1. Trying to combine TLS port-per-broker with TLS SNI.
2. Trying to combine TCP and TLS

and we currently don't prevent 1

Reply:
oh I see what you mean, shared ports should only be arriving with `requiresSni == true`. I think it's better for the conflict detector to be less aware of the conventions and check both cases.

Reply:
headache, so these cases shouldn't be possible right? If they are using port-per-broker they'll be exclusive regardless of TLS and fail earlier. If they're both using SNI then they should both be tls and requestSni==true.

So these are more like sanity checks.